packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -0,0 +1,203 @@
+          },
+        },
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      ignoreIntersections: false,
+      ignoreUnions: false,
+    },
+  ],
+  create(context, [{ ignoreIntersections, ignoreUnions }]) {
+    const sourceCode = context.getSourceCode();
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    function checkDuplicate(
+      node: TSESTree.TSIntersectionType | TSESTree.TSUnionType,
+    ): void {
+      const duplicateConstituents: {
+        duplicated: TSESTree.TypeNode;
+        duplicatePrevious: TSESTree.TypeNode;
+      }[] = [];
+      const cachedTypeMap: Map<TSESTree.TypeNode, Type> = new Map();
+      node.types.reduce<TSESTree.TypeNode[]>(
+        (uniqConstituents, constituentNode) => {
+          const duplicatedPreviousConstituentInAst = uniqConstituents.find(
+            ele => isSameAstNode(ele, constituentNode),
+          );
+          if (duplicatedPreviousConstituentInAst) {
+            duplicateConstituents.push({
+              duplicated: constituentNode,
+              duplicatePrevious: duplicatedPreviousConstituentInAst,
+            });
+            return uniqConstituents;
+          }
+          const constituentNodeType = checker.getTypeAtLocation(
+            parserServices.esTreeNodeToTSNodeMap.get(constituentNode),
+          );
+          const duplicatedPreviousConstituentInType = [
+            ...cachedTypeMap.entries(),
+          ].find(([, type]) => {
+            return type === constituentNodeType;
+          })?.[0];
+          if (duplicatedPreviousConstituentInType) {
+            duplicateConstituents.push({
+              duplicated: constituentNode,
+              duplicatePrevious: duplicatedPreviousConstituentInType,
+            });
+            return uniqConstituents;
+          }
+          cachedTypeMap.set(constituentNode, constituentNodeType);
+          return [...uniqConstituents, constituentNode];
+        },
+        [],
+      );
+      duplicateConstituents.forEach(duplicateConstituent => {
+        context.report({
+          data: {
+            duplicated: sourceCode.getText(duplicateConstituent.duplicated),
+            type:
+              node.type === AST_NODE_TYPES.TSIntersectionType
+                ? 'Intersection'
+                : 'Union',
+            previous: sourceCode.getText(
+              duplicateConstituent.duplicatePrevious,
+            ),
+          },
+          messageId: 'duplicate',
+          node,
+          fix: fixer => {
+            const beforeTokens = sourceCode.getTokensBefore(
+              duplicateConstituent.duplicated,
+            );
+            const afterTokens = sourceCode.getTokensAfter(
+              duplicateConstituent.duplicated,
+            );
+            const beforeUnionOrIntersectionToken = beforeTokens
+              .reverse()
+              .find(token => token.value === '|' || token.value === '&');
```

Comment:
i think this is a little faster
```suggestion
            const beforeTokens = sourceCode.getTokensBefore(
              duplicateConstituent.duplicated,
              { filter: token => token.value === '|' || token.value === '&' },
            );
            const beforeUnionOrIntersectionToken = beforeTokens[beforeTokens.length - 1];
```

Reply:
The way you taught me is more smart 👍